// @storybook/vue3/dist/index.js
import {
  entry_preview_exports,
  renderToCanvas,
  setup
} from "./_browser-chunks/chunk-GZHEVVQL.js";
import {
  __name
} from "./_browser-chunks/chunk-JFJ5UJ7Q.js";

// src/globals.ts
import { global } from "@storybook/global";
var { window: globalWindow } = global;
globalWindow.STORYBOOK_ENV = "vue3";
globalWindow.PLUGINS_SETUP_FUNCTIONS ||= /* @__PURE__ */ new Set();

// src/portable-stories.ts
import {
  composeStories as originalComposeStories,
  composeStory as originalComposeStory,
  setProjectAnnotations as originalSetProjectAnnotations,
  setDefaultProjectAnnotations
} from "storybook/preview-api";
import { h } from "vue";
function setProjectAnnotations(projectAnnotations) {
  setDefaultProjectAnnotations(vueProjectAnnotations);
  return originalSetProjectAnnotations(
    projectAnnotations
  );
}
__name(setProjectAnnotations, "setProjectAnnotations");
var vueProjectAnnotations = {
  ...entry_preview_exports,
  /** @deprecated */
  renderToCanvas: /* @__PURE__ */ __name((renderContext, canvasElement) => {
    if (renderContext.storyContext.testingLibraryRender == null) {
      return renderToCanvas(renderContext, canvasElement);
    }
    const {
      storyFn,
      storyContext: { testingLibraryRender: render }
    } = renderContext;
    const { unmount } = render(storyFn(), { container: canvasElement });
    return unmount;
  }, "renderToCanvas")
};
function composeStory(story, componentAnnotations, projectAnnotations, exportsName) {
  const composedStory = originalComposeStory(
    story,
    componentAnnotations,
    projectAnnotations,
    globalThis.globalProjectAnnotations ?? vueProjectAnnotations,
    exportsName
  );
  const renderable = /* @__PURE__ */ __name((...args) => h(composedStory(...args)), "renderable");
  Object.assign(renderable, composedStory);
  return renderable;
}
__name(composeStory, "composeStory");
function composeStories(csfExports, projectAnnotations) {
  const composedStories = originalComposeStories(csfExports, projectAnnotations, composeStory);
  return composedStories;
}
__name(composeStories, "composeStories");
export {
  composeStories,
  composeStory,
  setProjectAnnotations,
  setup,
  vueProjectAnnotations
};
