// @storybook/vue3-vite/dist/_node-chunks/chunk-AGEPCA7C.js
import CJS_COMPAT_NODE_URL_o0x6nzzn8bp from 'node:url';
import CJS_COMPAT_NODE_PATH_o0x6nzzn8bp from 'node:path';
import CJS_COMPAT_NODE_MODULE_o0x6nzzn8bp from "node:module";

var __filename = CJS_COMPAT_NODE_URL_o0x6nzzn8bp.fileURLToPath(import.meta.url);
var __dirname = CJS_COMPAT_NODE_PATH_o0x6nzzn8bp.dirname(__filename);
var require = CJS_COMPAT_NODE_MODULE_o0x6nzzn8bp.createRequire(import.meta.url);

// ------------------------------------------------------------
// end of CJS compatibility banner, injected by Storybook's esbuild configuration
// ------------------------------------------------------------
import {
  __name
} from "./chunk-CCRFNFVA.js";

// src/plugins/vue-template.ts
async function templateCompilation() {
  return {
    name: "storybook:vue-template-compilation",
    config: /* @__PURE__ */ __name(() => ({
      resolve: {
        alias: {
          vue: "vue/dist/vue.esm-bundler.js"
        }
      }
    }), "config")
  };
}
__name(templateCompilation, "templateCompilation");

export {
  templateCompilation
};
